pkg/maputil/maputil.go

```diff
@@ -60,32 +63,133 @@ func recursivelyStringifyMapKey(v interface{}) (interface{}, error) {
 	return casted_v, nil
 }
 
-func Set(m map[string]interface{}, key []string, value string) map[string]interface{} {
-	if len(key) == 0 {
-		panic(fmt.Errorf("bug: unexpected length of key: %d", len(key)))
-	}
+type arg interface {
+	getMap(map[string]interface{}) map[string]interface{}
+	assign(map[string]interface{}, string)
+}
 
-	k := key[0]
+type keyArg struct {
+	key string
+}
 
-	if len(key) == 1 {
-		m[k] = value
-		return m
+func (a keyArg) getMap(m map[string]interface{}) map[string]interface{} {
+	_, ok := m[a.key]
+	if !ok {
+		m[a.key] = map[string]interface{}{}
+	}
+	switch t := m[a.key].(type) {
+	case map[string]interface{}:
+		return t
+	default:
+		panic(fmt.Errorf("unexpected type: %v(%T)", t, t))
 	}
+}
 
-	remain := key[1:]
+func (a keyArg) assign(m map[string]interface{}, value string) {
+	m[a.key] = value
+}
+
+type indexedKeyArg struct {
+	key   string
+	index int
+}
 
-	nested, ok := m[k]
+func (a indexedKeyArg) getArray(m map[string]interface{}) []interface{} {
+	_, ok := m[a.key]
 	if !ok {
-		nested = map[string]interface{}{}
+		m[a.key] = make([]interface{}, a.index+1)
+	}
+	switch t := m[a.key].(type) {
+	case []interface{}:
+		if len(t) <= a.index {
+			t2 := make([]interface{}, a.index+1)
+			copy(t, t2)
+			t = t2
+		}
+		return t
+	default:
+		panic(fmt.Errorf("unexpected type: %v(%T)", t, t))
+	}
+}
+
+func (a indexedKeyArg) getMap(m map[string]interface{}) map[string]interface{} {
+	t := a.getArray(m)
+	if t[a.index] == nil {
+		t[a.index] = map[string]interface{}{}
 	}
-	switch t := nested.(type) {
+	switch t := t[a.index].(type) {
 	case map[string]interface{}:
-		nested = Set(t, remain, value)
+		return t
 	default:
 		panic(fmt.Errorf("unexpected type: %v(%T)", t, t))
 	}
+}
+
+func (a indexedKeyArg) assign(m map[string]interface{}, value string) {
+	t := a.getArray(m)
+	t[a.index] = value
+	m[a.key] = t
+}
+
+func getArg(key string) arg {
+	key = strings.ReplaceAll(key, "[", " ")
+	key = strings.ReplaceAll(key, "]", " ")
+	k := key
+	idx := 0
+
+	n, err := fmt.Sscanf(key, "%s %d", &k, &idx)
+
+	if n == 2 && err == nil {
+		return indexedKeyArg{
+			key:   k,
+			index: idx,
+		}
+	}
 
-	m[k] = nested
+	return keyArg{
+		key: key,
+	}
+}
 
-	return m
+func ParseKey(key string) []string {
+	r := []string{}
+	part := ""
+	escaped := false
+	for _, rune := range key {
+		split := false
+		switch {
+		case escaped == false && rune == '\\':
+			escaped = true
+			continue
+		case rune == '.':
+			split = escaped == false
+		}
+		escaped = false
+		if split {
+			r = append(r, part)
+			part = ""
+		} else {
+			part = part + string(rune)
+		}
+	}
+	if len(part) > 0 {
+		r = append(r, part)
+	}
+	return r
+}
+
+func Set(m map[string]interface{}, key []string, value string) {
+	if len(key) == 0 {
+		panic(fmt.Errorf("bug: unexpected length of key: %d", len(key)))
+	}
+
+	appender := getArg(key[0])
```

Comment:
I'd prefer this called something akin to `cursor` or `pointer` or `reference` so that it looks like:

```
cur := getCursor(key[0])

if len(key) == 1 {
    cursor.set(m, value)
} else {
    remain := key[1:]
    nested := cur.getMap(m)
    Set(nested, remain, value)
}
```

Reply:
done ;)

Reply:
I'm not fully sure, but perhaps this function can be shortened into this? 

```
func Set(m map[string]interface{}, key []string, value string) {
	if len(key) == 0 {
		panic(fmt.Errorf("bug: unexpected length of key: %d", len(key)))
	}
	
	for len(key) > 1 {
        	m, key = getCursor(key[0]).getMap(m), key[1:]
	}
	getCursor(key[0]).set(m, value)
}
```

Reply:
I think so, i thought that the recursion was not necessary but didn't take time to review it.
Sorry, i just pushed the change.